test/widgets/register_page_test.dart

```diff
@@ -0,0 +1,322 @@
+      await pumpMaterial(tester, const RegisterPage());
+
+      final RegisterFormState formState = tester.state(find.byType(RegisterForm));
+
+      final FormFieldState usernameField = tester.state(usernameFieldFinder);
+
+      // Not validated yet, so no error
+      expect(usernameField.hasError, isFalse);
+
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // Empty, so error
+      expect(usernameField.hasError, isTrue);
+
+      await tester.enterText(usernameFieldFinder, 'a');
+
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // Not empty, so no error
+      expect(usernameField.hasError, isFalse);
+
+      await tester.sendKeyDownEvent(LogicalKeyboardKey.backspace);
+
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // Empty, so error
+      expect(usernameField.hasError, isTrue);
+    });
+
+    testWidgets('Validates the password', (WidgetTester tester) async {
+      await pumpMaterial(tester, const RegisterPage());
+
+      final RegisterFormState formState = tester.state(find.byType(RegisterForm));
+
+      final FormFieldState passwordField = tester.state(passwordFieldFinder);
+
+      // Not validated yet, so no error
+      expect(passwordField.hasError, isFalse);
+
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // Empty, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, 'abcdefg');
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // Too short, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, 'abcdefgh');
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // No number, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, 'abcdefg0');
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // No capital letters, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, 'ABCDEFG0');
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // No lower letters, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, 'abcdEFG0');
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      // No special, so error
+      expect(passwordField.hasError, isTrue);
+
+      await tester.enterText(passwordFieldFinder, password);
+      formState.formKey.currentState!.validate();
+      await tester.pumpAndSettle();
+      expect(passwordField.hasError, isFalse);
+    });
+
+    testWidgets('Validates the password confirmation', (WidgetTester tester) async {
```

Comment:
what happens if they are the same but both  do not meet the requirements  just wondering?

Reply:
I guess both of them will show an error, but I don't think that's a relevant case right? It's covered by these two test cases.